SharedPackages/DataBrokerProtectionCore/Sources/DataBrokerProtectionCore/OptOutServices/EmailConfirmationDataService.swift

```diff
@@ -0,0 +1,153 @@
+
+        if featureFlagger.isEmailConfirmationDecouplingFeatureOn {
+            try database.saveOptOutEmailConfirmation(profileQueryId: profileQueryId,
+                                                     brokerId: dataBrokerId,
+                                                     extractedProfileId: extractedProfileId,
+                                                     generatedEmail: emailData.emailAddress,
+                                                     attemptID: attemptId.uuidString)
+        }
+
+        return emailData
+    }
+
+    public func getConfirmationLink(from email: String,
+                                    numberOfRetries: Int,
+                                    pollingInterval: TimeInterval,
+                                    attemptId: UUID,
+                                    shouldRunNextStep: @escaping () -> Bool) async throws -> URL {
+        try await emailServiceV0.getConfirmationLink(from: email,
+                                                     numberOfRetries: numberOfRetries,
+                                                     pollingInterval: pollingInterval,
+                                                     attemptId: attemptId,
+                                                     shouldRunNextStep: shouldRunNextStep)
+    }
+
+    public func checkForEmailConfirmationData() async throws {
+        let recordsAwaitingLink = try database.fetchOptOutEmailConfirmationsAwaitingLink()
+
+        var itemsToDelete: [EmailDataRequestItemV1] = []
+        for chunk in recordsAwaitingLink.chunks(ofCount: EmailServiceV1.Constants.maxBatchSize) {
+            let records = Array(chunk)
+            let response = try await emailServiceV1.fetchEmailData(items: records.toEmailDataRequestItems())
+
+            itemsToDelete.append(contentsOf: response.items.toEmailDataResponseItemsForDeletion())
+
+            for item in response.items {
+                switch item.status {
+                case .ready:
+                    if let record = records[email: item.email, attemptId: item.attemptId] {
+                        try database.updateOptOutEmailConfirmationLink(item.confirmationLink,
+                                                                       emailConfirmationLinkObtainedOnBEDate: item.linkObtainedOnBEDate,
+                                                                       profileQueryId: record.profileQueryId,
+                                                                       brokerId: record.brokerId,
+                                                                       extractedProfileId: record.extractedProfileId)
+                    }
+                case .pending:
+                    continue
+                case .unknown, .error:
+                    if let record = records[email: item.email, attemptId: item.attemptId] {
+                        try database.deleteOptOutEmailConfirmation(profileQueryId: record.profileQueryId,
```

Comment:
I assume this is deliberate as retrying logic will come later, but I imagine we don't always want to delete these in case of an error?

Reply:
`unknown` and `error` are considered unrecoverable, so there’s nothing we can do here: https://dub.duckduckgo.com/duckduckgo/dbp-api/pull/323/files#r96044

Reply:
Might be worth adding a comment in the code to explain it

Reply:
Funny that adding a comment to explain it revealed an edge case I forgot to handle: The corresponding opt-out job should be rescheduled as it’s now in .error state.